core/src/text.rs

```diff
@@ -0,0 +1,461 @@
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub enum Color {
+    DarkRed,
+    Red,
+    Gold,
+    Yellow,
+    DarkGreen,
+    Green,
+    Aqua,
+    DarkAqua,
+    DarkBlue,
+    Blue,
+    LightPurple,
+    DarkPurple,
+    White,
+    Gray,
+    DarkGray,
+    Black,
+}
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub enum Style {
+    Bold,
+    Italic,
+    Underlined,
+    Strikethrough,
+    Obfuscated,
+}
+
+// TODO: use instead of string as keybind
+pub enum KeyBind {
+    Attack,
+    UseItem,
+    Forward,
+    Left,
+    Back,
+    Right,
+    Jump,
+    Sneak,
+    Sprint,
+    Drop,
+    Iventory,
+    Chat,
+    ListPlayers,
+    PickBlock,
+    Command,
+    Screenshot,
+    Perspective,
+    MouseSmoothing,
+    Fullscreen,
+    SpectatorOutlines,
+    SaveToolbar,
+    LoadToolbar,
+    Advancements,
+    Hotbar1,
+    Hotbar2,
+    Hotbar3,
+    Hotbar4,
+    Hotbar5,
+    Hotbar6,
+    Hotbar7,
+    Hotbar8,
+    Hotbar9,
+    // Suport custom keybinds?
+    Custom(String),
+}
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+#[serde(tag = "action", content = "value")]
+// TODO: Accept any json primitive as string
+pub enum Click {
+    OpenUrl(String),
+    OpenFile(String),
+    RunCommand(String),
+    ChangePage(i32),
+    SuggestCommand(String),
+    CopyToClipboard(String),
+}
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+#[serde(tag = "action", content = "value")]
+// TODO: Accept any json primitive as string
+pub enum Hover {
+    #[serde(rename = "show_text")]
+    ShowText(Box<RawText>),
+    #[serde(rename = "show_item")]
+    // TODO: Item struct 
+    ShowItem(String),
+    #[serde(rename = "show_entity")] 
+    // TODO: Entity struct
+    ShowEntity(String),
+}
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum TextValue {
+    Text {
+        text: String,
```

Comment:
Could we use `Cow<'static, str>` to allow either static strings without allocation or heap-allocated `String`s? Functions which take an `&str` could then be switched to take `impl Into<Cow<'static, str>>`.